include/dice/template-library/channel.hpp

```diff
@@ -90,7 +91,13 @@ namespace dice::template_library {
 
             {
                 std::unique_lock lock{queue_mutex_};
-                queue_not_full_.wait(lock, [this]() noexcept { return queue_.size() < max_cap_; });
+                queue_not_full_.wait(lock, [this]() noexcept { return queue_.size() < max_cap_ || closed_.test(std::memory_order_acquire); });
+
+                if (queue_.size() >= max_cap_) [[unlikely]] {
+                    // wait was exited because closed_ was true (channel closed)
```

Comment:
I think this is not necessarily true. 
From how I understand the condition variables, the notification is not guaranteed to arrive immediately. 

So, a sequence of  
```
B: emplace() blocks on queue_not_full_.wait(lock, [this]() noexcept { return queue_.size() < max_cap_ || closed_.test(std::memory_order_acquire); });
A: close()
A: pop()
B: emplace() wakes up just now
```
would be valid, right?

In that case the element would still be emplaced. Is this intended behaviour? 

Otherwise, this could be prevented by setting max_cap_ to 0 in close().

Reply:
Right, so we can just check closed_ again instead. That should work

Reply:
I've noticed another thing: whenever we hold the lock its actually enough to check the closed_ flag with relaxed memory ordering, because its only ever modified under a lock. If we don't hold the lock we still need acquire I think